### PR TITLE
Scrubbing OneVsAll

### PR DESCRIPTION
**Pull request by singlis**
- Renamed OVA to OneVsAllTrainer
- Updates to comments and documentation
Related to #2619

### CODE REVIEW

**Comment by Ivanidzo4ka** [comment]
>             IEstimator<ISingleFeaturePredictionTransformer<ICalibrator>> calibrator = null,

@wschin do we have any plans to do something about this one?

---
Refers to: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs:597 in 24ecdcc. [](commit_id = 24ecdcc43ad9e687c91dcce637dfeda386097e30, deletion_comment = False)

**Comment by artidoro** [comment]
> // The .NET Foundation licenses this file to you under the MIT license.

Could you also rename the file?

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/MultiClass/Ova.cs:2 in 24ecdcc. [](commit_id = 24ecdcc43ad9e687c91dcce637dfeda386097e30, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs [hunk 1]
```diff
@@ -590,7 +590,7 @@ private static ICalibratorTrainer GetCalibratorTrainerOrThrow(IExceptionContext
         /// <param name="maxCalibrationExamples">Number of instances to train the calibrator.</param>
         /// <param name="useProbabilities">Use probabilities (vs. raw outputs) to identify top-score category.</param>
         /// <typeparam name="TModel">The type of the model. This type parameter will usually be inferred automatically from <paramref name="binaryEstimator"/>.</typeparam>
-        public static Ova OneVersusAll<TModel>(this MulticlassClassificationCatalog.MulticlassClassificationTrainers catalog,
+        public static OneVsAllTrainer OneVersusAll<TModel>(this MulticlassClassificationCatalog.MulticlassClassificationTrainers catalog,
             ITrainerEstimator<ISingleFeaturePredictionTransformer<TModel>, TModel> binaryEstimator,
             string labelColumnName = DefaultColumnNames.Label,
             bool imputeMissingLabelsAsNegative = false,
```
Comment (artidoro) [inline src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs]: Why do we need label and not features here?

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs]: I think this label column is purely for calibration. BinaryEstimator contains point on label and feature column.

---
In reply to: [260486559](https://github.com/dotnet/machinelearning/pull/2743#discussion_r260486559) [](ancestors = 260486559)

Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs]: Makes sense! Thank you @Ivanidzo4ka

**Comment by singlis** [comment]
> // The .NET Foundation licenses this file to you under the MIT license.

yes.

---
In reply to: [467612026](https://github.com/dotnet/machinelearning/pull/2743#issuecomment-467612026) [](ancestors = 467612026)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/MultiClass/Ova.cs:2 in 24ecdcc. [](commit_id = 24ecdcc43ad9e687c91dcce637dfeda386097e30, deletion_comment = False)

**Comment by wschin** [comment]
> IEstimator<ISingleFeaturePredictionTransformer<ICalibrator>> calibrator = null,
> 
> 
> @wschin do we have any plans to do something about this one?
> 
> Refers to: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs:597 in 24ecdcc. [](commit_id = 24ecdcc, deletion_comment = False)

This doesn't  look very good to me but my [assignment](https://github.com/dotnet/machinelearning/issues/2467) for fixing that is postponed.

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/MetaMulticlassTrainer.cs [hunk 2]
```diff
@@ -20,7 +20,7 @@ public abstract class MetaMulticlassTrainer<TTransformer, TModel> : ITrainerEsti
         where TTransformer : ISingleFeaturePredictionTransformer<TModel>
         where TModel : class
     {
-        public abstract class OptionsBase
+        internal abstract class OptionsBase
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/MetaMulticlassTrainer.cs]: Why do we internalize an `Options` of a trainer?

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs [hunk 3]
```diff
@@ -62,30 +62,27 @@ internal sealed class Options : OptionsBase
...
         }
 
         /// <summary>
-        /// Initializes a new instance of <see cref="Ova"/>.
+        /// Initializes a new instance of <see cref="OneVersusAllTrainer"/>.
         /// </summary>
         /// <param name="env">The <see cref="IHostEnvironment"/> instance.</param>
         /// <param name="binaryEstimator">An instance of a binary <see cref="ITrainerEstimator{TTransformer, TPredictor}"/> used as the base trainer.</param>
-        /// <param name="calibrator">The calibrator. If a calibrator is not explicitely provided, it will default to <see cref="PlattCalibratorTrainer"/></param>
+        /// <param name="calibrator">The calibrator. If a calibrator is not provided, it will default to <see cref="PlattCalibratorTrainer"/></param>
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs]: How does a naturally-calibrated model interact with this parameter?

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs [hunk 4]
```diff
@@ -62,30 +62,27 @@ internal sealed class Options : OptionsBase
...
         /// <param name="labelColumn">The name of the label colum.</param>
-        /// <param name="imputeMissingLabelsAsNegative">Whether to treat missing labels as having negative labels, instead of keeping them missing.</param>
+        /// <param name="imputeMissingLabelsAsNegative">If true will treat missing labels as negative labels.</param>
         /// <param name="maxCalibrationExamples">Number of instances to train the calibrator.</param>
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs]:
```suggestion
        /// <param name="maximumCalibrationExampleCount">Number of instances to train the calibrator.</param>
```
What's the indent scope of this PR? Inferred from its title, I assume that having better argument names is included. Please correct me if I am wrong.

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs [hunk 5]
```diff
@@ -62,30 +62,27 @@ internal sealed class Options : OptionsBase
...
         /// <param name="useProbabilities">Use probabilities (vs. raw outputs) to identify top-score category.</param>
-        internal Ova(IHostEnvironment env,
+        internal OneVersusAllTrainer(IHostEnvironment env,
             TScalarTrainer binaryEstimator,
             string labelColumn = DefaultColumnNames.Label,
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs]:
```suggestion
            string labelColumnName = DefaultColumnNames.Label,
```

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs [hunk 6]
```diff
@@ -62,30 +62,27 @@ internal sealed class Options : OptionsBase
...
 
         /// <summary>
-        /// Initializes a new instance of <see cref="Ova"/>.
+        /// Initializes a new instance of <see cref="OneVersusAllTrainer"/>.
         /// </summary>
         /// <param name="env">The <see cref="IHostEnvironment"/> instance.</param>
         /// <param name="binaryEstimator">An instance of a binary <see cref="ITrainerEstimator{TTransformer, TPredictor}"/> used as the base trainer.</param>
-        /// <param name="calibrator">The calibrator. If a calibrator is not explicitely provided, it will default to <see cref="PlattCalibratorTrainer"/></param>
+        /// <param name="calibrator">The calibrator. If a calibrator is not provided, it will default to <see cref="PlattCalibratorTrainer"/></param>
         /// <param name="labelColumn">The name of the label colum.</param>
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs]:
```suggestion
        /// <param name="labelColumnName">The name of the label colum.</param>
```

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/Pkpd.cs [hunk 7]
```diff
@@ -50,7 +50,7 @@ namespace Microsoft.ML.Trainers
     /// logistic regression is a more principled way to solve a multiclass problem, it
     /// requires that the learner store a lot more intermediate state in the form of
     /// L-BFGS history for all classes *simultaneously*, rather than just one-by-one
-    /// as would be needed for OVA.
+    /// as would be needed for OneVersusAll.
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/Pkpd.cs]: >OneVersusAll [](start = 31, length = 12)

Please use `see cref=` if you mean a class or just one-versus-all classification model otherwise.

**Comment by wschin** [comment]
>             int maxCalibrationExamples = 1000000000,

maxCalibrationExampleCount

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs:90 in 5093549. [](commit_id = 509354995726626f5bf32e7e737acd7c2aed758c, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs [hunk 8]
```diff
@@ -49,7 +49,7 @@ public sealed class Ova : MetaMulticlassTrainer<MulticlassPredictionTransformer<
         private readonly Options _options;
 
         /// <summary>
-        /// Options passed to OVA.
+        /// Options passed to OneVsAll.
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs]: Please do `see cref=` here.

**Comment by singlis** [comment]
>             int maxCalibrationExamples = 1000000000,

good catch - I think this should be maximumCalibrationExampleCount

---
In reply to: [468101653](https://github.com/dotnet/machinelearning/pull/2743#issuecomment-468101653) [](ancestors = 468101653)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/MultiClass/Ova.cs:90 in 5093549. [](commit_id = 509354995726626f5bf32e7e737acd7c2aed758c, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs (hunk 4 above)
Comment (singlis) [reply src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs]: Better argument names is in scope for this PR.

---
In reply to: [261016830](https://github.com/dotnet/machinelearning/pull/2743#discussion_r261016830) [](ancestors = 261016830)

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/MetaMulticlassTrainer.cs (hunk 2 above)
Comment (singlis) [reply src/Microsoft.ML.StandardLearners/Standard/MultiClass/MetaMulticlassTrainer.cs]: This is internal because the only class that derives from it, Options for pkpd and ova are both internal. 

The option classes are internal because there are no advanced options and all of the arguments are exposed in the argument-based constructor.


---
In reply to: [261015420](https://github.com/dotnet/machinelearning/pull/2743#discussion_r261015420) [](ancestors = 261015420)

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs (hunk 3 above)
Comment (singlis) [reply src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs]: I dont know for sure - I will posted an issue as a question. #2805

---
In reply to: [261015949](https://github.com/dotnet/machinelearning/pull/2743#discussion_r261015949) [](ancestors = 261015949)

Comment (singlis) [reply src/Microsoft.ML.StandardLearners/Standard/MultiClass/OneVersusAllTrainer.cs]: Also I think it would be worth updating this comment once I have more info. I won't that block checking this in for now though.

---
In reply to: [261430492](https://github.com/dotnet/machinelearning/pull/2743#discussion_r261430492) [](ancestors = 261430492,261015949)

File: build/Dependencies.props [hunk 9]
```diff
@@ -15,7 +15,7 @@
     <GoogleProtobufPackageVersion>3.5.1</GoogleProtobufPackageVersion>
     <LightGBMPackageVersion>2.2.3</LightGBMPackageVersion>
     <MicrosoftMLOnnxRuntimePackageVersion>0.2.1</MicrosoftMLOnnxRuntimePackageVersion>
-    <MlNetMklDepsPackageVersion>0.0.0.7</MlNetMklDepsPackageVersion>
+    <MlNetMklDepsPackageVersion>0.0.0.8</MlNetMklDepsPackageVersion>
```
Comment (singlis) [inline build/Dependencies.props]: ugh! I merged a branch. Will revert these changes..
